pipeline/bluesky_extract.py

```diff
@@ -0,0 +1,100 @@
+"""Extract script to connect to Bluesky Firehose and extract relevant data given topics"""
+
+import csv
+import ssl
+import json
+import re
+import certifi
+
+from atproto import CAR, models
+from atproto_client.models.utils import get_or_create
+from atproto_firehose import FirehoseSubscribeReposClient, parse_subscribe_repos_message
+
+
+class JSONExtra(json.JSONEncoder):
+    """Serializes raw objects (including CID-Content Identifier) as strings."""
+
+    def default(self, obj: bytes):
+        try:
+            return json.JSONEncoder.default(self, obj)
+        except (TypeError, ValueError, KeyError):
+            return repr(obj)
+
+
+def format_text(text: str) -> str:
+    """Removes extra lines and whitespaces from text"""
+    # Remove consecutive empty lines
+    text = re.sub(r'\n\s*\n', '\n', text)
+
+    text = text.strip()
+    # Replace multiple spaces with a single space
+    text = re.sub(r'\s+', ' ', text)
+
+    return text
+
+
+def extract_text_from_bytes(raw: bytes) -> str:
+    """Extracts text from a raw Bluesky post"""
+    try:
+        json_data = json.dumps(raw, cls=JSONExtra, indent=2)
+        parsed_json = json.loads(json_data)
+        text = parsed_json.get('text')
+        return format_text(text)
+    except (TypeError, AttributeError):
+        return None
+
+
+def get_firehose_data(message: bytes, topics: list[str],
+                      csv_writer: csv.writer, csvfile: csv.writer) -> None:
+    """Handles incoming messages, parses data, and writes to a CSV file based on dynamic topics."""
+    repo_commit = parse_subscribe_repos_message(message)
+    if not isinstance(repo_commit, models.ComAtprotoSyncSubscribeRepos.Commit):
+        return
+    car_file = CAR.from_bytes(repo_commit.blocks)
+    for operation in repo_commit.ops:
+        if operation.action in ["create"] and operation.cid:
+            raw_bytes = car_file.blocks.get(operation.cid)
+            processed_post = get_or_create(raw_bytes, strict=False)
+
+            if processed_post.py_type == "app.bsky.feed.post":
+                firehose_text = extract_text_from_bytes(raw_bytes)
+
+                for topic in topics:
+                    if topic in firehose_text:
```

Comment:
Would it make sense to check for the topic not including when it's part of another word e.g. looking for spy and catching crispy? Or is this something for the transform?

Reply:
That does make a lot of sense but maybe this is something that can be addressed in transform?